<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="front_camera_image_topic" default="/front_camera/color/image_raw"/>
  <arg name="right_camera_image_topic" default="/right_camera/color/image_raw"/>
  <arg name="back_camera_image_topic" default="/back_camera/color/image_raw"/>
  <arg name="left_camera_image_topic" default="/left_camera/color/image_raw"/>
  
  <arg name="yolo_model" default="yolov8n-pose.pt"/>
  <arg name="detection_topic" default="detection_result"/>
  <arg name="conf_thres" default="0.25"/>
  <arg name="iou_thres" default="0.45"/>
  <arg name="max_det" default="300"/>
  <arg name="tracker" default="bytetrack.yaml"/>
  <arg name="debug" default="true"/>
  <arg name="debug_conf" default="true"/>
  <arg name="debug_line_width" default="1"/>
  <arg name="debug_font_size" default="1"/>
  <arg name="debug_font" default="Arial.ttf"/>
  <arg name="debug_labels" default="true"/>
  <arg name="debug_boxes" default="true"/>

  <!-- Front Camera Node -->
  <node name="tracker_node_front" pkg="ultralytics_ros" type="tracker_node_gaz_multiple.py">
    <param name="image_topic" value="$(arg front_camera_image_topic)"/>
    <param name="debug_image_topic" value="debug_image_front"/>
    <param name="yolo_model" value="$(arg yolo_model)"/>
    <param name="detection_topic" value="$(arg detection_topic)"/>
    <param name="conf_thres" value="$(arg conf_thres)"/>
    <param name="iou_thres" value="$(arg iou_thres)"/>
    <param name="max_det" value="$(arg max_det)"/>
    <param name="tracker" value="$(arg tracker)"/>
    <param name="debug" value="$(arg debug)"/>
    <param name="debug_conf" value="$(arg debug_conf)"/>
    <param name="debug_line_width" value="$(arg debug_line_width)"/>
    <param name="debug_font_size" value="$(arg debug_font_size)"/>
    <param name="debug_font" value="$(arg debug_font)"/>
    <param name="debug_labels" value="$(arg debug_labels)"/>
    <param name="debug_boxes" value="$(arg debug_boxes)"/>
  </node>
  
  <!-- Right Camera Node -->
  <node name="tracker_node_right" pkg="ultralytics_ros" type="tracker_node_gaz_multiple.py">
    <param name="image_topic" value="$(arg right_camera_image_topic)"/>
    <param name="debug_image_topic" value="debug_image_right"/>
    <param name="yolo_model" value="$(arg yolo_model)"/>
    <param name="detection_topic" value="$(arg detection_topic)"/>
    <param name="conf_thres" value="$(arg conf_thres)"/>
    <param name="iou_thres" value="$(arg iou_thres)"/>
    <param name="max_det" value="$(arg max_det)"/>
    <param name="tracker" value="$(arg tracker)"/>
    <param name="debug" value="$(arg debug)"/>
    <param name="debug_conf" value="$(arg debug_conf)"/>
    <param name="debug_line_width" value="$(arg debug_line_width)"/>
    <param name="debug_font_size" value="$(arg debug_font_size)"/>
    <param name="debug_font" value="$(arg debug_font)"/>
    <param name="debug_labels" value="$(arg debug_labels)"/>
    <param name="debug_boxes" value="$(arg debug_boxes)"/>
  </node>

  <!-- Back Camera Node -->
  <node name="tracker_node_back" pkg="ultralytics_ros" type="tracker_node_gaz_multiple.py">
    <param name="image_topic" value="$(arg back_camera_image_topic)"/>
    <param name="debug_image_topic" value="debug_image_back"/>
    <param name="yolo_model" value="$(arg yolo_model)"/>
    <param name="detection_topic" value="$(arg detection_topic)"/>
    <param name="conf_thres" value="$(arg conf_thres)"/>
    <param name="iou_thres" value="$(arg iou_thres)"/>
    <param name="max_det" value="$(arg max_det)"/>
    <param name="tracker" value="$(arg tracker)"/>
    <param name="debug" value="$(arg debug)"/>
    <param name="debug_conf" value="$(arg debug_conf)"/>
    <param name="debug_line_width" value="$(arg debug_line_width)"/>
    <param name="debug_font_size" value="$(arg debug_font_size)"/>
    <param name="debug_font" value="$(arg debug_font)"/>
    <param name="debug_labels" value="$(arg debug_labels)"/>
    <param name="debug_boxes" value="$(arg debug_boxes)"/>
  </node>

  
  <!-- Left Camera Node -->
 <node name="tracker_node_left" pkg="ultralytics_ros" type="tracker_node_gaz_multiple.py">
    <param name="image_topic" value="$(arg left_camera_image_topic)"/>
    <param name="debug_image_topic" value="debug_image_left"/>
    <param name="yolo_model" value="$(arg yolo_model)"/>
    <param name="detection_topic" value="$(arg detection_topic)"/>
    <param name="conf_thres" value="$(arg conf_thres)"/>
    <param name="iou_thres" value="$(arg iou_thres)"/>
    <param name="max_det" value="$(arg max_det)"/>
    <param name="tracker" value="$(arg tracker)"/>
    <param name="debug" value="$(arg debug)"/>
    <param name="debug_conf" value="$(arg debug_conf)"/>
    <param name="debug_line_width" value="$(arg debug_line_width)"/>
    <param name="debug_font_size" value="$(arg debug_font_size)"/>
    <param name="debug_font" value="$(arg debug_font)"/>
    <param name="debug_labels" value="$(arg debug_labels)"/>
    <param name="debug_boxes" value="$(arg debug_boxes)"/>
  </node>

  <!-- Image Combiner Node -->
  <node name="image_combiner" pkg="ultralytics_ros" type="image_combiner.py">
    <param name="debug_image_front_topic" value="debug_image_front"/>
    <param name="debug_image_right_topic" value="debug_image_right"/> 
    <param name="debug_image_back_topic" value="debug_image_back"/>
    <param name="debug_image_left_topic" value="debug_image_left"/>
    <param name="debug_image_combined_topic" value="debug_image_combined"/>
  </node>

  <!-- Image View Node -->
  <node if="$(arg debug)" name="image_view" pkg="image_view" type="image_view" output="screen">
    <param name="autosize" value="false"/>
    <remap from="image" to="debug_image_combined"/>
  </node>
</launch>





